internal/mapdecode/hooks.go

```diff
@@ -0,0 +1,180 @@
+	}
+}
+
+// Wraps a reflectHook to support pointers in either direction (source and
+// destination).
+func supportPointers(hook reflectHook) (outputHook reflectHook) {
+	outputHook = func(from, to reflect.Type, data reflect.Value) (reflect.Value, error) {
+		// Get rid of pointers in either direction. This lets us parse **foo if we
+		// know how to parse foo.
+		switch {
+		case from == to:
+			return data, nil
+		case from.Kind() == reflect.Ptr: // *foo => bar
+			// Decoding a pointer type to a non-pointer type. Dereference if
+			// non-nil, use zero-value otherwise.
+			from = from.Elem()
+			if data.IsNil() {
+				data = reflect.Zero(from)
+			} else {
+				data = data.Elem()
+			}
+			return outputHook(from, to, data)
+		case to.Kind() == reflect.Ptr: // foo => *bar
+			// Decoding a non-pointer type to a pointer. Decode as usual and take
+			// a pointer to the result.
+			out, err := outputHook(from, to.Elem(), data)
+			if err != nil {
+				return out, err
+			}
+
+			// If we didn't know what to do with the input, the returned value
+			// will just be the data as-is and it won't have the correct type.
+			if !out.Type().AssignableTo(to.Elem()) {
+				return data, nil
+			}
+
+			result := reflect.New(to.Elem())
+			result.Elem().Set(out)
+			return result, nil
+		}
+
+		return hook(from, to, data)
+	}
+	return
+}
+
+// Builds a reflectHook which unmarshals types using the given unmarshaling
+// scheme. See the unmarshaler type for more information.
+func unmarshalerHook(opts *options) reflectHook {
+	return func(from, to reflect.Type, data reflect.Value) (reflect.Value, error) {
+		if !reflect.PtrTo(to).Implements(opts.Unmarshaler.Interface) {
+			return data, nil
+		}
+
+		// The following lines are roughly equivalent to,
+		// 	value := new(foo)
+		// 	err := value.Decode(...)
+		// 	return *value, err
+		value := reflect.New(to)
+		err := opts.Unmarshaler.Unmarshal(value, decodeFrom(opts, data.Interface()))
+		if err != nil {
+			err = fmt.Errorf("could not decode %v from %v: %v", to, from, err)
+		}
+		return value.Elem(), err
+	}
+}
+
+// A reflectHook which decodes strings into time.Durations.
+func durationHook(from, to reflect.Type, data reflect.Value) (reflect.Value, error) {
+	if from.Kind() != reflect.String || to != _typeOfDuration {
+		return data, nil
+	}
+
+	d, err := time.ParseDuration(data.String())
+	return reflect.ValueOf(d), err
+}
+
+// stringToPrimitivesHook is a reflectHook which decodes strings into
+// primitives.
+//
+// Integers are parsed in base 10.
+func strconvHook(from, to reflect.Type, data reflect.Value) (reflect.Value, error) {
+	// TODO(abg): If necessary, we can add parsing of hexadecimal and octal
+	// integers too.
+
+	if from.Kind() != reflect.String {
+		return data, nil
+	}
+
+	s := data.String()
+	switch to.Kind() {
+	case reflect.Bool:
+		b, err := strconv.ParseBool(s)
+		return reflect.ValueOf(b), err
+	case reflect.Float32, reflect.Float64:
+		f, err := strconv.ParseFloat(s, to.Bits())
```

Comment:
Don't you need to check for overflows for numerics? I think they'll silently become zeros.